Add infercnv plots to QC

sjspielman opened the pull request:
Closes #999
Closes #1000 

This PR adds the next two sections to the inferCNV report; each was fairly small so I went ahead and did them together.
The first new section is UMAP colored by total CNV, and the second new section is ridge plots of total CNV. For the ridge plot across cell types, I decided to mirror the cell type report section and show the same lumped cell types there. Let me know if you agree with that decision! In order to do this, I had to move the `lump_wrap_celltypes()` into the report functions file instead of the celltype functions file so it was available to the ifnerCNV chld report.

Here is an attached rendered report (again no cell type sections for faster local development). Note that these plots are _not compelling_ biologically since the processed SCE I used to generate them was produced from an SCE I ran through while testing and used that exciting new inferCNV argument to limit the number of cells per group for runtime, so the CNV estimates are bad (note this is _not_ related to NA situation I'm looking into; indeed I need to do a full rerun of infercnv over there to track things down...).
[infercnv-main_qc_report.html.zip](https://github.com/user-attachments/files/22392084/infercnv-main_qc_report.html.zip)

sjspielman committed:
update and move the lump_wrap function into report_functions.rmd so the infercnv qc can use it

sjspielman committed:
add UMAP section

sjspielman committed:
add ridge plot section

sjspielman committed:
Merge branch 'development' into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
Merge sjspielman/998-initialize-infercnv-report

sjspielman committed:
use eval=has_infercnv_results chunks

sjspielman committed:
Merge branch 'sjspielman/998-initialize-infercnv-report' into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
variable name is now infercnv_total_cnv

sjspielman committed:
Merge branch 'sjspielman/999-1000-infercnv-qc-plots' of github.com:AlexsLemonade/scpca-nf into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
Merge branch 'development' into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
fewer chunks now

sjspielman committed:
instead of another argument to the function, just rename for consistency in the infercnv report

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

allyhawkins reviewed (commented):
This looks pretty good, I just had some suggestions for clean up and then one note about making sure we don't have an error when a cell type only has 2 cells.

sjspielman committed:
Merge branch 'development' into sjspielman/999-1000-infercnv-qc-plots

sjspielman commented on templates/qc_report/infercnv_qc.rmd:
```diff
@@ -95,3 +95,92 @@ For more information on interpreting this heatmap, please refer to the [`inferCN
+  dplyr::mutate(
+    is_infercnv_reference = ifelse(is_infercnv_reference, "Reference cells", "Query cells"),
+    is_infercnv_reference = glue::glue("{is_infercnv_reference} (N={n})")
+  )
+
+total_cnv_filtered_df <- total_cnv_df |>
+  # rename for consistency with cell type report section
```
I was confused why I had this code there but remembered, it was left over from some initial cell filtering to remove cell types with <3 which I should not have deleted because it's important for the ridge plots! I'll address this alongside your other comment about <3 cells.

sjspielman replied:
Woops didn't finish that thought! 
The reason for a separate filtered object is, we want to remove the <3 cell types from the data going into the cell type ridge plot, but _not_ going into the plot faceted by cell _group_.

sjspielman committed:
Update templates/qc_report/infercnv_qc.rmd

Co-authored-by: Ally Hawkins <54039191+allyhawkins@users.noreply.github.com>

sjspielman committed:
Merge branch 'development' into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
Merge branch 'sjspielman/999-1000-infercnv-qc-plots' of github.com:AlexsLemonade/scpca-nf into sjspielman/999-1000-infercnv-qc-plots

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sjspielman committed:
Merge branch 'sjspielman/999-1000-infercnv-qc-plots' of github.com:AlexsLemonade/scpca-nf into sjspielman/999-1000-infercnv-qc-plots

sjspielman committed:
formally define and use a consensus palette, and fix df bug to use the validation palette df

sjspielman committed:
update wrangling to filter to 3 cells and use the existing palette

sjspielman committed:
Merge branch 'development' into sjspielman/999-1000-infercnv-qc-plots

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sjspielman committed:
add eval=has_umap

sjspielman committed:
Merge branch 'sjspielman/999-1000-infercnv-qc-plots' of github.com:AlexsLemonade/scpca-nf into sjspielman/999-1000-infercnv-qc-plots

sjspielman commented:
I ended up trying to get a little fancy with it. Let me know if it's too fancy, or if you can think of anything that might break this approach!

 I modified the celltypes report to define a consensus palette used in the UMAP separately, and I updated the `faceted_umap()` function to use a palette if provided. Then, I use this palette in the inferCNV report. This means the color <-> celltype relationship is preserved in the inferCNV report since it relies on the separate mapping between hex colors and labels.

While I was in the celltype report, I caught a bug too - wrong variable name was being used for the marker palette df. 

Here's a _complete_ rendered report (aka including cell type sections):
[infercnv.html.zip](https://github.com/user-attachments/files/22412150/infercnv.html.zip)

allyhawkins commented on templates/qc_report/infercnv_qc.rmd:
```diff
@@ -95,3 +95,117 @@ For more information on interpreting this heatmap, please refer to the [`inferCN
+The first ridge plot shows the CNV distributions between cell type groups: Those included in the `inferCNV` normal reference, versus all other query cells.
+
+The second ridge plot shows the CNV distributions across consensus cell types, showing only the top seven most common cell types.
+Note that any cell types with fewer than 3 cells are not shown in this plot; if one of the most common cell types had fewer than 3 cells, the colors in this plot may differ from those shown in the cell type annotation QC report section.
```
```suggestion
Note that any cell types with fewer than 3 cells are not shown in this plot.
```
I think you fixed this by creating the palette right? So you shouldn't need that statement.

sjspielman replied:
💯 , thanks!

allyhawkins reviewed (approved):
LGTM, just one minor change to make.

sjspielman committed:
Update templates/qc_report/infercnv_qc.rmd

Co-authored-by: Ally Hawkins <54039191+allyhawkins@users.noreply.github.com>
